lqngen/lqn2lqx.1
.\" -*- nroff -*-
.TH lqn2lqx 1 ""  "5.13"
.\" $Id: lqn2lqx.1 16195 2022-12-23 22:03:52Z greg $
.\"
.\" --------------------------------
.SH "NAME"
lqn2lqx \- generate layered queueing network models from an existing model file.
.SH "SYNOPSIS"
.br
.B lqngen
[\fIOPTION \&.\|.\|.\fP]
[
FILE
]
.SH "DESCRIPTION"
\fBlqn2lqx\fP is used to convert an LQN input file to use either SPEX or LQX to control execution.
All constant input parameters are converted to variables initialized with the values from the original
input file.  Existing variables within the input file are not
modified.  Note however, that variables initialized in an XML input file with LQX are \fBnot\fP initialized
in the converted output file because the LQX program must be executed in order to do so.  To convert input formats without the conversion of parameters to
variables, use \fBlqn2ps\fP(1) with \fB--format\fP=\fIxml\fP (or \fIlqn\fP or 
\fIjson\fP).
.PP
\fBlqn2lqx\fP reads its input from \fIfilename\fP, specified at the
command line if present, or from the standard input otherwise.  Output
for an input file \fIfilename\fP specified on the command line will be
placed in the file \fIfilename.ext\fP, where \fI.ext\fP is \fIxlqn\fP
for SPEX conversion, \fIlqnx\fPfor LQX conversion and \fIjson\fP
for JSON conversion.
If the output file name is the same as the input file name, 
the output is written back to the original file name.
The original file is renamed to \fIfilename.ext~\fP
Output can be directed to a new file with the \fB\-\-output\fP option.
If several input files are given, then each is treated as a separate model and
output will be placed in separate output files.
.SH "OPTIONS"
.TP
\fB\-N\fR, \fB\-\-experiments=\fIARG\fR
Generate \fIARG\fP experiments within one model file. The experiments
will be run using either SPEX (for LQN input), or LQX (for XML input).
This option will turn on either \fB\-\-spex-output\fP, or \fB\-\-lqx-output\fP
depending on the output format.
.TP
\fB\-S\fR, \fB\-\-sensitivity=\fIARG\fR
Create a factorial expermiment varying all non-reference task service times by a factor of plus or minus \fIARG\fP.  This option precludes the use of \fB\-N\fP.
.TP
\fB\-O\fR, \fB\-\-format=\fIARG\fR
Set the output file format, \fIARG\fP, to: \fIxml\fP, \fIlqn\fP.
.TP
\fB\-H\fR, \fB\-\-help\fR
Print out a brief help summary and exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Print the version number of lqn2lqx.
.TP
\fB\-c\fR, \fB\-\-customers[=\fIARG]\fR
Set the \fImean\fP number of customers at each of the client reference tasks to \fIARG\fP.
\fIARG\fP can be take the form of \fB\-\-customers\fP=\fImean\fP, 
\fB\-\-customers\fP=\fIa:b\fP where \fIa\fP and \fIb\fP are parameters
for the default discreet distribution, or \fB\-\-customers\fP=\fIdistribution:a[:b]\fP, where 
\fIdistribution\fP is a \fBdiscreet\fP distribution, and \fIa[:b]\fP are parameters to the distribution.
By default, a \fIpoisson\fP distribution is used.
.TP
\fB\-g\fR, \fB\-\-share[=\fIARG]\fR
Set the share of the first group running on the processor to \fIARG\fP.
The share for a group is based on the Beta distribution.
.TP
\fB\-d\fR, \fB\-\-delay-server[=\fIARG]\fR
Set the probability that a processor is a delay server to \fIARG\fP.
Tasks are not affected.
All processors that have constant multiplicities are eligible for conversion to infinite servers.
Processors whose multiplicity is set using a variable in the input file are never changed.
.TP
\fB\-i\fR, \fB\-\-infinite-server[=\fIARG]\fR
Set the probability that a server task is infinite to \fIARG\fP.
Processors are not affected.
All tasks that have constant multiplicities are eligible for conversion to infinite servers.
Tasks whose multiplicity is set using a variable in the input file are never changed.
.TP
\fB\-p\fR, \fB\-\-processor-multiplicity[=\fIARG]\fR
Set the \fImean\fP processor multiplicity for each processor to \fIARG\fP.
\fIARG\fP can be take the form of \fB\-\-processor-multiplicity\fP=\fImean\fP, 
\fB\-\-processor-multiplicity\fP=\fIa:b\fP where \fIa\fP and \fIb\fP are parameters
for the default discreet distribution, or \fB\-\-processor-multiplicity\fP=\fIdistribution:a[:b]\fP, where 
\fIdistribution\fP is a \fBdiscreet\fP distribution, and \fIa[:b]\fP are parameters to the distribution.
By default, a \fIpoisson\fP distribution is used.
.TP
\fB\-s\fR, \fB\-\-service-time[=\fIARG]\fR
Set the \fImean\fP phase service time to \fIARG\fP.
\fIARG\fP can be take the form of \fB\-\-service-time\fP=\fImean\fP, 
\fB\-\-service-time\fP=\fIa:b\fP where \fIa\fP and \fIb\fP are parameters
for the default continuous distribution, or \fB\-\-service-time\fP=\fIdistribution:a[:b]\fP, where 
\fIdistribution\fP is a \fBcontinuous\fP distribution, and \fIa[:b]\fP are parameters to the distribution.
By default, a \fIgamma\fP distribution is used.
.TP
\fB\-t\fR, \fB\-\-task-multiplicity[=\fIARG]\fR
Set the \fImean\fP task multiplicity for each task to \fIARG\fP.
\fIARG\fP can be take the form of \fB\-\-task-multiplicity\fP=\fImean\fP, 
\fB\-\-task-multiplicity\fP=\fIa:b\fP where \fIa\fP and \fIb\fP are parameters
for the default discreet distribution, or \fB\-\-task-multiplicity\fP=\fIdistribution:a[:b]\fP, where 
\fIdistribution\fP is a \fBdiscreet\fP distribution, and \fIa[:b]\fP are parameters to the distribution.
By default, a \fIpoisson\fP distribution is used.
Tasks in the input file which are infinite servers may be converted to fixed-rate or multi-servers.
.TP
\fB\-y\fR, \fB\-\-request-rate[=\fIARG]\fR
Set the mean rendezous (synchronous call) rate to \fIARG\fP.  
\fIARG\fP can be take the form of \fB\-\-request-rate\fP=\fImean\fP, 
\fB\-\-request-rate\fP=\fIa:b\fP where \fIa\fP and \fIb\fP are parameters
for the default continuous distribution, or \fB\-\-request-rate\fP=\fIdistribution:a[:b]\fP, where 
\fIdistribution\fP is a \fBcontinuous\fP distribution, and \fIa[:b]\fP are parameters to the distribution.
By default, a \fIgamma\fP distribution is used.
.TP
\fB\-z\fR, \fB\-\-think-time[=\fIARG]\fR
Set the mean think time at reference tasks to \fIARG\fP.
\fIARG\fP can be take the form of \fB\-\-think-time\fP=\fImean\fP, 
\fB\-\-think-time\fP=\fIa:b\fP where \fIa\fP and \fIb\fP are parameters
for the default continuous distribution, or \fB\-\-think-time\fP=\fIdistribution:a[:b]\fP, where 
\fIdistribution\fP is a \fBcontinuous\fP distribution, and \fIa[:b]\fP are parameters to the distribution.
By default, a \fIgamma\fP distribution is used.
.TP
\fB\-o\fR, \fB\-\-output=\fIARG\fR
Redirect output to the file named \fIARG\fP.  If \fIARG\fP is \fB-\fP,
output is redirected to \fIstdout\fP.  This option cannot be used with \fB\-\-models\fP.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Verbose output. List the actual number of Clients, Server Tasks, Processors, Entries and Calls created.
.TP
\fB\-\-json-output\fR
Output the input model in JavaScript Object Notation (JSON).
.TP
\fB\-\-xml-output\fR
Output the input model in eXtensible Markup Language (XML).
.TP
\fB\-\-lqx-output\fR
Generate LQX control code for XML output files.  Variables will be
created for all parameters set by the options above and will be
initialized using the current random number generator.  If
\fB\-\-experiments\fP=\fIARG\fP is also used, loop code will be
produced to generate \fIARG\fP runs.  This option will enable XML output.
.TP
\fB\-\-spex-output\fR
Generate SPEX control code for LQN output files.  Variables will be
created for all parameters set by the options above and will be
initialized using the current random number generator.  If
\fB\-\-experiments\fP=\fIARG\fP is also used, loop code will be
produce to generate \fIARG\fP runs.  This option will enable LQN
output.
.TP
\fB\-\-seed=\fIARG\fR
Set the seed value for the random number generator to \fIARG\fP.
.TP
\fB\-\-beta=\fIARG\fR
The BETA distribution is only used for choosing the share of a group when using a processor
using Completely Fair Scheduling.
Set the beta argument of the distribution to \fIARG\fP.  The alpha
argument is set based on the value of the group "share" (set using \fB\-g\fP\fIn\fP).
.TP
\fB\-\-binomial\fR
Use a Binomial distribution for all subsequent options that use a
discreet random variable generator.
.TP
\fB\-\-constant\fR
Use \fIconstant\fP values for all subsequent parameters.
.TP
\fB\-\-normal=\fIARG\fR
Set the continuous random variable generator to use a Normal
distribution with a standard deviation of \fIARG\fP for any subsequent
flag that uses a continuous distribution.
.TP
\fB\-\-gamma=\fIARG\fR
Set the continuous random variable generator to use a Gamma
distribution with a shape parameter of \fIARG\fP  for any subsequent
flag that uses a continuous distribution.  Integer values of
\fIARG\fP will generate random numbers with an Erlang distribution.  A
shape value of 1 will generate random numbers with an Exponential
distribution.
.TP
\fB\-\-poisson\fR
Use a Poisson distribution for all subsequent options that use a
discreet random variable generator.  The distribution is shifted right by one so that
the lower bound on generated values is always one.  Mean values are adjusted accordingly.
.TP
\fB\-\-uniform=\fIARG\fR
Use a uniform distribution for all subsequent options that use a
discreet or continuous random variable generator.  The spread of the
distribution is set to \fIARG\fP although this value will be
ignored if a mean value is set.  The lower bound for continuous distributions
is zero.  For discreet distributions, the lower bound is one.  The upper bound
for either distribution is set to twice the mean plus the lower bound.
.TP
\fB\-\-total-customers=\fIARG\fR
Set the total number of customers to \fIARG\fP.
\fIARG\fP can be take the form of \fB\-\-total-customers\fP=\fImean\fP, 
\fB\-\-total-customers\fP=\fIa:b\fP where \fIa\fP and \fIb\fP are parameters
for the default discreet distribution, or \fB\-\-total-customers\fP=\fIdistribution:a[:b]\fP, where 
\fIdistribution\fP is a \fBdiscreet\fP distribution, and \fIa[:b]\fP are parameters to the distribution.
By default, a \fIpoisson\fP distribution is used.
This option cannot be used with \fB\-\-customers\fP=\fIn\fP.
.TP
\fB\-\-no-variables\fR
Do not convert constant values for customers, processor and task multiplicities, service and think times and request
rates into variables.  Observation variables are not affected.  
.TP
\fB\-\-no-observation\fR
Do not insert any LQX code or spex observation variables.
.TP
\fB\-\-no-customers\fR
The default operation is to convert all constants in the input file into variables.
Do not convert constant parameters for reference task multiplicites (customers)
to variables.
.TP
\fB\-\-no-processor-multiplicity\fR
The default operation is to convert all constants in the input file into variables.
Do not convert constant parameters for processor multiplicities to variables.
.TP
\fB\-\-no-service-time\fR
The default operation is to convert all constants in the input file into variables.
Do not convert constant parameters for phase service times to variables.
.TP
\fB\-\-no-task-multiplicity\fR
The default operation is to convert all constants in the input file into variables.
Do not convert constant parameters for task multiplicities to variables.
.TP
\fB\-\-no-request-rate\fR
The default operation is to convert all constants in the input file into variables.
Do not convert constant parameters for request rates to variables.
.TP
\fB\-\-no-think-time\fR
The default operation is to convert all constants in the input file into variables.
Do not convert constant parameters for reference task (customer) think times to variables.
.TP
\fB\-\-[no-]input-parameters\fR
When generating SPEX or LQX output, either include or not include in the output the value of the parameters that change
as a result of executing the program.  Constant parameters are not included.
The default is to include the parameters.
.TP
\fB\-\-[no-]throughput\fR
Insert LQX code or SPEX observation variables to output task throughput for all ``intersting'' tasks.
Interesting tasks are those which might have contention present.
The default is to insert observation variables.
.TP
\fB\-\-[no-]residence-time\fR
Insert LQX code or SPEX observation variables to output entry service (residence) time.
The default is to insert observation variables.
.TP
\fB\-\-[no-]waiting-time\fR
Insert LQX code or SPEX observation variables to either observe or not observe the waiting time between phases and entries.
The default is to observe waiting time.
.TP
\fB\-\-[no-]utilization\fR
Insert LQX code or SPEX observation variables to either observe or not observe processor utilization for all ``interesting'' processors.
Interesting processors are those which might have contention present.
The default is to observe processor utilization.
.TP
\fB\-\-[no-]mva-steps\fR
Insert LQX code or SPEX observation variables to output the number of calls to step().
The default is to not insert observation variables.
.TP
\fB\-\-[no-]mva-waits\fR
Insert LQX code or SPEX observation variables to output the number of calls to wait().
The default is to not insert observation variables.
.TP
\fB\-\-[no-]iterations\fR
Insert LQX code or SPEX observation variables to output the number of solver iterations.
The default is to not insert observation variables.
.TP
\fB\-\-[no-]elapsed-time\fR
Insert LQX code or SPEX observation variables to output the solver's ELAPSED time.
The default is to not insert observation variables.
.TP
\fB\-\-[no-]user-cpu-time\fR
Insert LQX code or SPEX observation variables to output the solver's USER CPU time.
The default is to not insert observation variables.
.TP
\fB\-\-[no-]system-cpu-time\fR
Insert LQX code or SPEX observation variables to output the solver's SYSTEM CPU time.
The default is to not insert observation variables.
.TP
\fB\-\-[no-]pragma[=\fIARG]\fR
Insert a pragma \fIARG\fP into all translated
This option can be repeated for multiple pragmas.
If the pragma was already present in the input file, it is reset to the new value.
\fB\-\-no\-pragma\fP will remove all existing pragmas.
.TP
\fB\-\-comment=\fIARG\fR
Set the model comment to \fIARG\fP.  By default, the comment is set to
the command line options of the invocation of \fBlqn2lqx\fP.
.TP
\fB\-\-convergence-value=\fIARG\fR
Set the model convergence limit to \fIARG\fP.  By default, the
convergence limit is set to 0.00001.
.TP
\fB\-\-under-relaxation=\fIARG\fR
Set the model under-relaxation to \fIARG\fP.  By default, the
under-relaxation is set to 0.9.
.TP
\fB\-\-iteration-limit=\fIARG\fR
Set the model iteration limit to \fIARG\fP.  By default, the iteration
limit is set to 50.  For models with many layers, this  value should
be higher.
.TP
\fB\-\-[no-]annotate\fR
Annotate LQN-type input files with syntax help.  This option has no effect for
XML or JSON output.  The default is to not annotate LQN model files.
.TP
\fB\-\-manual-page\fR
Generate the manual page and send the output to \fIstdout\fP.

.SH "SEE ALSO"
\fBlqn2lqx\fP(1), \fBlqns\fP(1), \fBlqsim\fP(1), \fBlqn2ps\fP(1)

.SH "EXAMPLES"
To convert an existing model file to SPEX:
.sp
.ti 0.75i
\f(CWlqn2lqx model.lqn\fP
.sp
Note that the output will be in a file named \fImodel.xlqn\fP.
.sp
To convert an existing model file to SPEX with running two experiments varying service time:
.sp
.ti 0.75i
\f(CWlqn2lqx -N2 -s2 model.lqn\fP
.sp
Note that the output will be in a file named \fImodel.xlqn\fP.
.sp
To convert an existing model file to LQX, varying the service time at all entries by 1.5x:
.sp
.ti 0.75i
\f(CWlqn2lqx --lqx-output --sensitivity=1.5 --no-customers --no-request-rate model.lqn\fP
.sp
.
.sp
To convert an existing model file to SPEX, and converting all serving tasks to infinite servers:
.sp
.ti 0.75i
\f(CWlqn2lqx --no-conversion -i1 model.lqn\fP
.sp
.
.sp
To convert an existing model file to SPEX, and converting all serving tasks to fixed rate servers:
.sp
.ti 0.75i
\f(CWlqn2lqx --no-conversion --constant -t1 model.lqn\fP
.sp
.
.sp


NEWS
* Bugs fixed in 5.27
192 	libmva 	 	MVA::marginalProbabilities2 mis-dimensioned. 
373 	lqns 	 	Processor waiting times at Rolia multiserver make no sense. 
380 	lqns 	 	lqns hangs with Conway multiserver. 
382 	lqn2xxx 	lqn2ps -Oout doesn't print processors. 
383 	lqiolib 	<pre> precedences can accept more than one activity in XML (JSON?) input. 
385 	petrisrv 	V5 petrisrvn: faulty models causing aborts.
386 	lqn2csv 	getopt_long headaches. 
387 	srvndiff 	File glob not working. 
389 	srvndiff 	C:\msys64\mingw32\bin\srvndiff.exe: model-001.lqxo:12: error: unclosed token 
390 	lqn2xxx 	lqn2xxx: command not found. 
391 	lqsim 		lqsim/regression/Makefile.am, 5*-replication. 
392 	petrisrv 	Calls are not being constructed
393 	lqiolib 	Save (and print) marginal queue probabilities.
395 	petrisrv 	V5 petrisrvn: faulty models causing aborts or failing.
397 	lqiolib 	Add qnap2 parser for BCMP models.
398 	lqiolib 	Refactor LQX math handing from ExternalVariables to LQX Expressions.
399 	lqiolib 	WhatIf construction is bogus.
400 	lqn2xxx 	Converting from json to lqn fails on Windows.
401 	libmva 		libc++abi: terminating with uncaught exception of type std::invalid_argument: default-output
402 	miscella 	Mac M1 issues. 	2022-11-30

* Bugs fixed in 5.26
365 	lqns	 	The warning for zero service time should be suppressed for reference tasks with think time. 
369 	petrisrv 	Segmentation violation with --overtaking 
370 	lqiolib 	Activity parameters probabilities and rates are constants. 
379 	lqns	 	V5 Regression. 	
372 	lqiolib 	Result variables can be reused. 
374 	lqns 		Asynchronous Cycles #1 	
375 	lqns 		Asynchronous Cycles #1 	
376 	lqn2xxx 	Overloaded infinite servers with open arrivals... 
378 	lqn2xxx 	Activity throughput not initialized. 

* Bugs fixed in 5.25
338 	lqns 		Queuing time not equal to zero when NCust = NServers for Zhou multiserver. 
352 	lqn2xxx 	--no-xxxx doesn't work. 
353 	lqns 		Assertion failed: (ix && ix <= sz), function operator[], file vector.h, line 174. 
357 	para_lib 	parasol/include/sem*caphore.h conflicts with /usr/include/semaphore.h 
360 	lqngen	 	lqngen -L<n> doesn't work. 
361 	lqn2xxx 	EXC_BAD_ACCESS (code=EXC_I386_GPFLT) LQIO::DOM::Document::~Document(this=0x0000000100a04080) at dom_document.cpp:123:3 
362 	lqn2csv 	terminate called after throwing an instance of 'std::runtime_error' 
363 	lqn2csv 	vector::_M_range_check: __n (which is 18446744073709551615) >= this->size() (which is 3) 

* Bugs fixed in 5.24
321 	lqiolib 	Runtime Exception Occured: Object of type phase does not have property service_time 
328 	spex 		Using $1, $2... as result variables fails. 
339 	lqiolib 	--reload-lqx segvs or generally is borked. 
340 	lqiolib 	loadResults not implemented in Json_Document. 
341 	lqn2xxx 	lqn2svg -Q1 -o 01-sanity-queue.svg -Zprune 01-sanity.in aborted. 
342 	qnsolver 	Assertion failed: (L[n][0] == nullptr), function reset, file mva.cc, line 353. 
344 	lqiolib 	Automatic variable creation generates duplicate symbols for 2-class models. 
345 	libmva 		No results for NCust = [X,0] 
346 	libmva 		class(station("m"),"k").throughput fails. 
347 	lqns 		Changes needed to -zgenerate 
351 	lqn2xxx 	EMF output borked. 

* Bugs fixes in 5.23
300 	lqiolib 	Station results not computed for single class models. 	
324 	lqns 		Runtime Exception Occured: Batch_Model::addToSubmodel: bad submodel for task t1 
329 	qnsolver 	uncaught exception of type std::out_of_range: map::at: key not found 
331 	lqn2xxx 	Rendering issues with activities. 
332 	lqsim 		Lqsim CFS: 71-fair, 72-fair 
271 	lqn2xxx 	Group by regex may or may not work. 
321 	lqiolib 	Runtime Exception Occured: Object of type phase does not have property service_time 
328 	spex 		Using $1, $2... as result variables fails. 
209 	lqns 		Odd throughput results 

* Bugs fixed in 5.22
314 	lqns	 	Add -Pmultiserver=zhou 
316 	lqns 	 	Allow zero's for rendezvous/send-no-reply requests 
317 	petrisrv 	SIGSEGV, Segmentation fault. in LQIO::DOM::Entity::getSchedulingType (this=0x0) at dom_entity.cpp:55 	
319 	qnsolver 	Can't build qnsolver without expat. 
322 	qnsolver 	Add -Pmultiserver 
325 	lqn2xxx 	Graphical output is borked. 

* Bugs fixed in 5.21
179     lqns            Replication does not take into consideration interlock.
228     lqns            Casale_1 model. Simulation and analytic solver differences.
305     lqns            44-activities.in: fails in old syntax, passes in XML.
311     lqiolib         Inline expressions not working (non loop code).
315     lqns            Phase think time not working.
259     lqns            Failures in faulty models.
299     lqns            Deprecate existing replication code.

* Bugs fixed in 5.20
311     lqiolib         Inline expressions not working (non loop code).
259     lqns            Failures in faulty models.
179     lqns            Replication does not take into consideration interlock.
228     lqns            Casale_1 model. Simulation and analytic solver differences.
305     lqns            44-activities.in: fails in old syntax, passes in XML.
299     lqns            Deprecate existing replication code.

* Bugs fixed in 5.19
294     lqns            07-sanity 
208     lqiolib         Add --gnuplot output. 
290     lqx             lqx parser doesn't like -1... sometimes. 
286     lqiolib         GNUPlot enhancements.   
293     lqns            Phase think time not being used by analytic solver. 
196     lqx             Print only array key or value as a comma separated list.
211     lqn2xxx         SIGSEGV Layer::transmorgrify(LQIO::DOM::Document*, Processor*&, Task*&)
266     lqn2xxx         LQIO::should_implement  
269     lqn2xxx         EXC_BAD_ACCESS lqn2svg`Layer::aggregate(this=0x0000000101009c70) at layer.cc:514:17
272     lqns            failure with telephone switch model 
274     lqn2xxx         lqn2lqn: error: Invalid rate for processor "server": 0.5 < 1.
276     lqiolib         lqn2xml -Olqx does not output LQX. 

lqngen/lqn2lqx.cc
/* lqngen.cc	-- Greg Franks Thu Jul 29 2004
 * Convert input files to lqnx with lqx parameterization.
 *
 * $Id: lqngen.cc 11811 2013-11-08 22:10:39Z greg $
 */

#include <errno.h>
#include <fstream>
#include <cstring>
#include <lqio/commandline.h>
#include <lqio/filename.h>
#include "lqngen.h"
#include "generate.h"

int
lqn2lqx( int argc, char **argv )
{
    Flags::lqx_output 	  = true;	
    Flags::xml_output 	  = true;	
    Flags::lqn2lqx    	  = true;
    Flags::annotate_input = false;

    unsigned int errorCode;
    extern char *optarg;
    char * output_file_name = 0;

    Generate::__service_time 		= new RV::Deterministic( 1.0 );
    Generate::__think_time 		= new RV::Deterministic( 0.0 );
    Generate::__forwarding_probability  = new RV::Deterministic( 0.0 );
    Generate::__rendezvous_rate 	= new RV::Deterministic( 1.0 );
    Generate::__send_no_reply_rate 	= new RV::Deterministic( 0.0 );
    Generate::__customers_per_client 	= new RV::Deterministic( 1.0 );
    Generate::__task_multiplicity 	= new RV::Deterministic( 1.0 );
    Generate::__processor_multiplicity  = new RV::Deterministic( 1.0 );
    Generate::__probability_second_phase= new RV::Deterministic( 0.0 );
    Generate::__number_of_entries	= new RV::Deterministic( 1.0 );

    static string opts = "";
#if HAVE_GETOPT_H
    static std::vector<struct option> longopts;
    makeopts( opts, longopts );
#if __cplusplus < 201103L
    LQIO::CommandLine command_line( opts, &longopts.front() );
#else
    LQIO::CommandLine command_line( opts, longopts.data() );
#endif
#else
    makeopts( opts );
    LQIO::CommandLine command_line( opts );
#endif

    command_line = io_vars.lq_toolname;

    optarg = 0;
    for ( ;; ) {
	char * endptr = 0;
#if HAVE_GETOPT_LONG
#if __cplusplus < 201103L
	const int c = getopt_long( argc, argv, opts.c_str(), &longopts.front(), NULL );
#else
	const int c = getopt_long( argc, argv, opts.c_str(), longopts.data(), NULL );
#endif
#else	
	const int c = getopt( argc, argv, opts.c_str() );
#endif
	if ( c == EOF) break;

        command_line.append( c, optarg );

	switch( c ) {
	case 'o':
	    output_file_name = optarg;
	    break;

	default:
	    if ( !common_arg( c, optarg, &endptr ) ) {
		usage();
		exit( 1 );
	    }
	    break;
	}

	if ( endptr != 0 && *endptr != '\0' ) {
	    fprintf( stderr, "%s: Invalid argumement to -%c: %s\n", io_vars.lq_toolname, c, optarg );
	    exit( 1 );
	}
    }

    io_vars.lq_command_line = command_line.c_str();

    if ( Flags::xml_output ) {
	Flags::lqx_output = true;	
	Flags::spex_output = false;
    } else {
	Flags::lqx_output = false;	
	Flags::spex_output = true;
    }

    if ( Flags::number_of_runs > 1 && Flags::sensitivity > 0 ) {
	fprintf( stderr, "%s: --experiments=%d and --sensitivity=%g are mutually exclusive.\n", io_vars.lq_toolname, Flags::number_of_runs, Flags::sensitivity );
	exit( 1 );
    }

    if ( optind == argc ) {
	LQIO::DOM::Document* document = LQIO::DOM::Document::load( "-", LQIO::DOM::Document::AUTOMATIC_INPUT, "", &io_vars, errorCode, false );
	if ( document ) {
	    Generate aModel( document, Flags::number_of_runs );
	    aModel.reparameterize();
	    if ( output_file_name && strcmp( output_file_name, "-" ) != 0 ) {
		std::ofstream output;
		output.open( output_file_name, ios::out|ios::binary );
		if ( !output ) {
		    cerr << "Cannot open output file " << output_file_name << " - " << strerror( errno );
		} else {
		    output << aModel;
		    output.close();
		}
	    } else {
		cout << aModel;
	    }
	}

    } else {
	for ( ;optind < argc; ++optind ) {
	    /* We need an Expat Document because we have to export our DOM.  The Xerces DOM isn't there */
	    LQIO::DOM::Document* document = LQIO::DOM::Document::load( argv[optind], LQIO::DOM::Document::AUTOMATIC_INPUT, "", &io_vars, errorCode, false );
	    if ( !document ) {
		continue;
	    } 
	    Generate aModel( document, Flags::number_of_runs );
	    aModel.reparameterize();
	    
	    if ( output_file_name && strcmp( output_file_name, "-" ) == 0 ) {
		cout << aModel;
	    } else {
		LQIO::Filename filename;
		if ( output_file_name ) {
		    filename.generate( output_file_name );
		} else {
		    const char * suffix = 0;
		    if ( Flags::xml_output ) suffix = "lqnx";
		    else suffix = "lqn";
		    filename.generate( argv[optind], suffix );
		}
		filename.backup();		/* Overwriting input file. -- back up */

		ofstream output;
		output.open( filename(), ios::out );
		if ( !output ) {
		    cerr << io_vars.lq_toolname << ": Cannot open output file " << filename() << " - "
			 << strerror( errno ) << endl;
		    exit ( 1 );
		}
		output << aModel;
		output.close();
	    }
	}
    }
    return 0;
}



